### PR TITLE
Image uploads

### PR DESCRIPTION
**Pull request by or-else**
I've implemented the same functionality as in web and Android for uploading large image and sending the URL.
I've refactored the progress indicator: it's now a bar instead of a circle. It works OK for both attachments and images. The circle looked odd with images. 

There is at least one unfixed bug (unrelated to uploads): when message sending fails and I delete the failed unsent message, the app tries to send {del} request to the server, fails, and that {del}-fail cycle is stuck forever.

### CODE REVIEW

File: Tinodios/AttributedStringFormatter.swift [hunk 1]
```diff
@@ -387,21 +389,35 @@ class AttributedStringFormatter: DraftyFormatter {
...
+
+                second.endEditing()
+                attributed.append(second)
             }
 
             if isValid {
                 // Insert linebreak then a clickable [↓ save] line
                 attributed.append(NSAttributedString(string: "\u{2009}\n", attributes: [NSAttributedString.Key.font : baseFont]))
 
-                // \u{2009} because iOS is buggy and bugs go unfixed for years.
```
Comment (aforge) [inline Tinodios/AttributedStringFormatter.swift:397]: Why remove this comment?
The context may be helpful in the future.

File: Tinodios/LargeFileHelper.swift [hunk 2]
```diff
@@ -1,64 +1,94 @@
...
+    public func finished(msg: ServerMessage?, err: Error?) {
+        self.finalCb?(msg, err)
+        self.finalCb = nil
+    }
 }
 
-class LargeFileHelper: NSObject {
-    static let kBoundary = "*****\(Date().millisecondsSince1970)*****"
+public class LargeFileHelper: NSObject {
+    static let kBoundary = "*****\(Int64((Date().timeIntervalSince1970 * 1000.0).rounded()))*****"
```
Comment (aforge) [inline Tinodios/LargeFileHelper.swift:60]: `millisecondsSince1970` is defined as `(timeIntervalSince1970 * 1000.0).rounded()`.

File: Tinodios/LargeFileHelper.swift [hunk 3]
```diff
@@ -1,64 +1,94 @@
...
+    convenience init(with tinode: Tinode) {
         let config = URLSessionConfiguration.background(withIdentifier: Bundle.main.bundleIdentifier!)
-        self.init(config: config)
+        self.init(with: tinode, config: config)
     }
+
     public static func addCommonHeaders(to request: inout URLRequest, using tinode: Tinode) {
         request.addValue(tinode.apiKey, forHTTPHeaderField: "X-Tinode-APIKey")
-        request.addValue("Token \(tinode.authToken!)", forHTTPHeaderField: "X-Tinode-Auth")
+        if tinode.isConnectionAuthenticated {
```
Comment (aforge) [inline Tinodios/LargeFileHelper.swift:80]: Won't the upload fail if the connection isn't authenticated?
Maybe return a `bool` from this function and handle it in the caller?

File: Tinodios/AttributedStringFormatter.swift (hunk 1 above)
Comment (or-else) [reply Tinodios/AttributedStringFormatter.swift:397]: I moved it up to the top of the code block.

File: Tinodios/LargeFileHelper.swift (hunk 2 above)
Comment (or-else) [reply Tinodios/LargeFileHelper.swift:60]: I tried to move LargeFileHelper to the SDK, failed, moved it back and this was lost in the process. Sorry. I put it back.

File: Tinodios/LargeFileHelper.swift (hunk 3 above)
Comment (or-else) [reply Tinodios/LargeFileHelper.swift:80]: I think it's not appropriate to return anything from this call. The check `isConnectionAuthenticated` is available to the caller.

**Comment by or-else** [comment]
It looks like the bug with deletion of failed messages is also present in Android.

**Comment by aforge** [comment]
> It looks like the bug with deletion of failed messages is also present in Android. I'm going to try to fix them both before release.

Sounds good to me.

**Review by aforge — APPROVED** [review]

**Pull request merged**